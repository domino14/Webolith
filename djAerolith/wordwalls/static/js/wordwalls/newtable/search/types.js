import searcher from '../../gen/rpc/wordsearcher/searcher_pb';

const pbsrConditions = searcher.SearchRequest.Condition;

const SearchTypesInputs = {
  TWO_NUMBERS: 1,
  ONE_NUMBER: 2,
  ONE_STRING: 3,
  SELECT: 4,
};

const SearchTypesEnum = {
  PROBABILITY: pbsrConditions.PROBABILITY_RANGE,
  LENGTH: pbsrConditions.LENGTH,
  TAGS: pbsrConditions.HAS_TAGS,
  POINTS: pbsrConditions.POINT_VALUE,
  NUM_ANAGRAMS: pbsrConditions.NUMBER_OF_ANAGRAMS,
  NUM_VOWELS: pbsrConditions.NUMBER_OF_VOWELS,
  FIXED_LENGTH: pbsrConditions.SINGLE_VALUE_LENGTH,
  NUM_TWO_BLANKS: pbsrConditions.NUM_TWO_BLANKS,
  MAX_SOLUTIONS: pbsrConditions.MAX_SOLUTIONS,
  NOT_IN_LEXICON: pbsrConditions.NOT_IN_LEXICON,
  PROBABILITY_LIMIT: pbsrConditions.PROBABILITY_LIMIT,
  MATCHING_ANAGRAM: pbsrConditions.MATCHING_ANAGRAM,
  DIFFICULTY_RANGE: pbsrConditions.DIFFICULTY_RANGE,
  /**
   * The inputs won't allow user to go beyond minAllowed and maxAllowed.
   * defaultMin and defaultMax are the values that show up when the
   * input is first placed on the screen.
   */
  properties: {
    [pbsrConditions.PROBABILITY_RANGE]: {
      code: pbsrConditions.PROBABILITY_RANGE,
      displayName: 'Probability Range',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 1,
      defaultMax: 100,
      minAllowed: 1,
      maxAllowed: 70000, // XXX should be subject to change by lexicon
      description: `Probability range lets you pick the top alphagrams by
      raw probability of drawing these tiles. Note: probability assumes
      there are two blanks in the bag, and alphagrams with identical
      probabilities will still have different (but consecutive) numbers.`,
    },
    [pbsrConditions.LENGTH]: {
      code: pbsrConditions.LENGTH,
      displayName: 'Word Length',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 2,
      defaultMax: 15,
      minAllowed: 2,
      maxAllowed: 15,
      description: 'You can filter by length of word with this option.',
    },
    [pbsrConditions.HAS_TAGS]: {
      code: pbsrConditions.HAS_TAGS,
      displayName: 'Has Tags',
      default: '',
      inputType: SearchTypesInputs.ONE_STRING,
      description: 'This is an unreleased feature, hold tight :)',
    },
    [pbsrConditions.POINT_VALUE]: {
      code: pbsrConditions.POINT_VALUE,
      displayName: 'Point Value',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 2,
      defaultMax: 30,
      minAllowed: 2,
      maxAllowed: 150, // for ZZZZZZZZZZZZZZZ
      description: `Filter by point value of the word. Letter values are taken
      from a certain crossword game HINT HINT.`,
    },
    [pbsrConditions.NUMBER_OF_ANAGRAMS]: {
      code: pbsrConditions.NUMBER_OF_ANAGRAMS,
      displayName: 'Number of Anagrams',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 1,
      defaultMax: 15,
      minAllowed: 1,
      maxAllowed: 99,
      description: `The number of total anagrams of this word, including
      the word itself.`,
    },
    [pbsrConditions.NUMBER_OF_VOWELS]: {
      code: pbsrConditions.NUMBER_OF_VOWELS,
      displayName: 'Number of Vowels',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 0,
      defaultMax: 15,
      minAllowed: 0,
      maxAllowed: 15,
      description: 'The number of vowels in this word',
    },
    [pbsrConditions.SINGLE_VALUE_LENGTH]: {
      code: pbsrConditions.SINGLE_VALUE_LENGTH,
      displayName: 'Word Length',
      inputType: SearchTypesInputs.ONE_NUMBER,
      default: 5,
      minAllowed: 5,
      maxAllowed: 10,
      description: 'You can filter by length of word with this option.',
    },
    [pbsrConditions.NUM_TWO_BLANKS]: {
      code: pbsrConditions.NUM_TWO_BLANKS,
      displayName: 'Number of 2-blank questions',
      inputType: SearchTypesInputs.ONE_NUMBER,
      default: 4,
      minAllowed: 1,
      maxAllowed: 50,
      description: 'How many questions with 2-blanks to generate in total.',
    },
    [pbsrConditions.MAX_SOLUTIONS]: {
      code: pbsrConditions.MAX_SOLUTIONS,
      displayName: 'Maximum number of anagrams',
      inputType: SearchTypesInputs.ONE_NUMBER,
      default: 5,
      minAllowed: 1,
      maxAllowed: 200,
      description: `The maximum number of anagrams that a question can have.
      No questions will be generated that have more than this number of anagrams.`,
    },
    [pbsrConditions.NOT_IN_LEXICON]: {
      code: pbsrConditions.NOT_IN_LEXICON,
      displayName: 'Not in lexicon',
      inputType: SearchTypesInputs.SELECT,
      default: 'other_english',
      options: [
        ['other_english', 'Other English-language lexicon'],
        ['update', 'Last version of this lexicon'],
      ],
      description: `Only looks for words that are NOT in the selected lexicon.
      The "other English-language lexicon" refers to CSW19 if you are currently
      using the NWL20 lexicon, and NWL20 if you are currently using the
      CSW19 lexicon. The "last version" is only CSW15 if you are using the CSW19
      lexicon. NWL20 has no new words compared to NWL18.`,
    },
    [pbsrConditions.PROBABILITY_LIMIT]: {
      code: pbsrConditions.PROBABILITY_LIMIT,
      displayName: 'Probability limit',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 1,
      defaultMax: 100,
      minAllowed: 1,
      maxAllowed: 70000,
      description: `Probability limit is similar to probability range, but
      it is applied after all other searches. For example, you can select the
      top 100 8-letter words with 5 vowels, and you will get 100 questions. If
      you used the "Probability range" option, though, you would only get any
      questions that were already in the 1-100 range by intrinsic probability,
      prior to applying any filters.`,
    },
    [pbsrConditions.MATCHING_ANAGRAM]: {
      code: pbsrConditions.MATCHING_ANAGRAM,
      displayName: 'Anagram match',
      inputType: SearchTypesInputs.ONE_STRING,
      default: 'AEINST??',
      description: `Provide an initial list of words to search with an anagram
      search. You can use up to 8 blank characters (use a ? character to
      represent a blank).`,
    },
    [pbsrConditions.DIFFICULTY_RANGE]: {
      code: pbsrConditions.DIFFICULTY_RANGE,
      displayName: 'Difficulty range',
      inputType: SearchTypesInputs.TWO_NUMBERS,
      defaultMin: 1,
      defaultMax: 100,
      minAllowed: 1,
      maxAllowed: 100,
      description: `Difficulty range is only applicable to 7 and 8 letter words
      in NWL20 and CSW19 currently. These words were
      split into 100 equal-size groups by difficulty, ranging from 1 (easiest)
      to 100 (hardest).`,
    },
  },
};

const SearchTypesOrder = [
  SearchTypesEnum.LENGTH,
  SearchTypesEnum.PROBABILITY,
  SearchTypesEnum.POINTS,
  SearchTypesEnum.MATCHING_ANAGRAM,
  SearchTypesEnum.NUM_ANAGRAMS,
  SearchTypesEnum.NUM_VOWELS,
  SearchTypesEnum.NOT_IN_LEXICON,
  SearchTypesEnum.FIXED_LENGTH,
  SearchTypesEnum.DIFFICULTY_RANGE,
  SearchTypesEnum.TAGS,
  SearchTypesEnum.NUM_TWO_BLANKS,
  SearchTypesEnum.MAX_SOLUTIONS,
  SearchTypesEnum.PROBABILITY_LIMIT,
];

function searchCriteriaOptions(allowedSearchTypes) {
  return SearchTypesOrder
    .filter((t) => allowedSearchTypes.has(t))
    .map((el) => ({
      value: String(el),
      displayValue: SearchTypesEnum.properties[el].displayName,
    }));
}

class SearchCriterion {
  constructor(searchType, options) {
    this.searchType = searchType;
    this.options = options;
  }

  inputType() {
    return SearchTypesEnum.properties[this.searchType].inputType;
  }

  deepCopy() {
    return new SearchCriterion(this.searchType, {
      ...this.options,
    });
  }

  /**
   * Convert this to an object that the backend would understand.
   */
  toJSObj() {
    const obj = {
      searchType: SearchTypesEnum.properties[this.searchType].code,
      ...this.options,
    };
    return obj;
  }

  /**
   * Set the option to the passed-in value. This function takes care
   * of converting the optionValue to the proper type, based on the
   * optionName
   * @param {string} optionName The name of the option in the options dict
   * @param {any} optionValue The value for the option
   */
  setOption(optionName, optionValue) {
    const valueModifier = (val) => {
      switch (this.inputType()) {
        case SearchTypesInputs.ONE_NUMBER:
        case SearchTypesInputs.TWO_NUMBERS:
          return parseInt(val, 10) || 0;
        case SearchTypesInputs.ONE_STRING:
          return val.trim();
        case SearchTypesInputs.SELECT:
          return val;
        default:
          throw new Error('Unsupported option name');
      }
    };
    this.options[optionName] = valueModifier(optionValue);
  }

  setOptions(options) {
    Object.keys(options).forEach((key) => this.setOption(key, options[key]));
  }

  resetSearchType(searchType) {
    this.searchType = searchType;
    // Reset the values.
    switch (SearchTypesEnum.properties[searchType].inputType) {
      case SearchTypesInputs.TWO_NUMBERS:
        this.setOptions({
          minValue: SearchTypesEnum.properties[searchType].defaultMin,
          maxValue: SearchTypesEnum.properties[searchType].defaultMax,
        });
        break;
      case SearchTypesInputs.ONE_NUMBER:
      case SearchTypesInputs.ONE_STRING:
      case SearchTypesInputs.SELECT:
        this.setOptions({
          value: SearchTypesEnum.properties[searchType].default,
        });
        break;
      default:
        break;
    }
  }
}

/**
 * Given a list of word search criteria, figure out the next one to add,
 * based on the order.
 * @param {Array.<Object>} wordSearchCriteria
 * @param {Set} allowedSearchTypes
 * @return {SearchCriterion?} A new search type criterion, or null.
 */
function searchCriterionToAdd(wordSearchCriteria, allowedSearchTypes) {
  const stmap = {};
  wordSearchCriteria.forEach((criteria) => {
    // These are enum integers.
    stmap[criteria.searchType] = true;
  });
  // Find the first search type that's not in the map.
  const newtypeId = SearchTypesOrder
    .filter((t) => allowedSearchTypes.has(t))
    .find((tid) => stmap[tid] == null);

  if (!newtypeId) {
    return null;
  }

  const typeProps = SearchTypesEnum.properties[newtypeId];
  switch (typeProps.inputType) {
    case SearchTypesInputs.ONE_STRING:
    case SearchTypesInputs.ONE_NUMBER:
    case SearchTypesInputs.SELECT:
      return new SearchCriterion(newtypeId, {
        value: SearchTypesEnum.properties[newtypeId].default,
      });
    case SearchTypesInputs.TWO_NUMBERS:
      return new SearchCriterion(newtypeId, {
        minValue: SearchTypesEnum.properties[newtypeId].defaultMin,
        maxValue: SearchTypesEnum.properties[newtypeId].defaultMax,
      });
    default:
      break;
  }
  return null;
}

export {
  SearchTypesEnum,
  SearchTypesOrder,
  searchCriterionToAdd,
  searchCriteriaOptions,
  SearchCriterion,
  SearchTypesInputs,
};
