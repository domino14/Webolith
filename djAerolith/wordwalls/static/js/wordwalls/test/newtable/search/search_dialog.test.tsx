import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import sinon from 'sinon';

import { SearchTypesEnum } from 'wordvaultapp/search/types';
import SearchDialogContainer from '../../../newtable/search/dialog_container';

interface MockAPI {
  call: sinon.SinonStub;
}

interface SearchDialogProps {
  lexicon: number;
  desiredTime: number;
  questionsPerRound: number;
  notifyError: (err: string) => void;
  redirectUrl: () => void;
  tablenum: number;
  onLoadNewList: () => void;
  disabled: boolean;
  showSpinner: () => void;
  hideSpinner: () => void;
  api: MockAPI | (() => void);
}

const props: SearchDialogProps = {
  lexicon: 3,
  desiredTime: 100,
  questionsPerRound: 50,
  // eslint-disable-next-line no-console
  notifyError: (err: string) => { console.log('Mock notifyError: ', err); },
  redirectUrl: () => { },
  tablenum: 12,
  onLoadNewList: () => { },
  disabled: false,
  showSpinner: () => { },
  hideSpinner: () => { },
  api: () => { },
};

const apiSpy = sinon.stub().returns({
  then: sinon.stub().returns({
    catch: sinon.stub().returns({
      finally: sinon.stub(),
    }),
  }),
});

const propsWithSpy: SearchDialogProps = {
  ...props,
  api: {
    call: apiSpy,
  },
};

describe('<SearchDialogContainer />', () => {
  describe('Static Analysis', () => {
    it('renders correctly', () => {
      const container = render(<SearchDialogContainer {...props} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Data', () => {
    it('submits expected parameters to back-end', () => {
      render(<SearchDialogContainer {...propsWithSpy} />);

      // Simulate the click.
      fireEvent.click(screen.getByText('Play!'));

      sinon.assert.calledWith(apiSpy, '/wordwalls/api/new_search/', {
        lexicon: 3,
        searchCriteria: [{
          searchType: 1,
          minValue: 7,
          maxValue: 7,
        }, {
          searchType: 2,
          minValue: 1,
          maxValue: 200,
        }],
        desiredTime: 100,
        questionsPerRound: 50,
        tablenum: 12,
      });
    });
  });

  describe('Interactions', () => {
    it('adds a search row when button is clicked', () => {
      render(<SearchDialogContainer {...props} />);
      // The very first button should be a `+`
      fireEvent.click(screen.getAllByRole('button')[0]);
      expect(screen.getAllByRole('combobox').length).toBe(3);
    });
  });

  it('removes a search row when button is clicked', () => {
    render(<SearchDialogContainer {...props} />);
    // Click the second button (minus)
    fireEvent.click(screen.getAllByRole('button')[1]);
    expect(screen.getAllByRole('combobox').length).toBe(1);
  });

  it('submits more search parameters to api after button is clicked', () => {
    render(<SearchDialogContainer {...propsWithSpy} />);
    fireEvent.click(screen.getAllByRole('button')[0]);
    fireEvent.click(screen.getByText('Play!'));

    sinon.assert.calledWith(apiSpy, '/wordwalls/api/new_search/', {
      lexicon: 3,
      searchCriteria: [{
        searchType: 1,
        minValue: 7,
        maxValue: 7,
      }, {
        searchType: 2,
        minValue: 1,
        maxValue: 200,
      }, {
        searchType: 8,
        minValue: 2,
        maxValue: 30,
      }],
      desiredTime: 100,
      questionsPerRound: 50,
      tablenum: 12,
    });
  });

  it('does not ever add single-value word length selector', () => {
    render(<SearchDialogContainer {...props} />);
    for (let i = 0; i < 15; i += 1) {
      // Click 'add' a whole bunch of times.
      fireEvent.click(screen.getAllByRole('button')[0]);
    }
    expect(screen.getAllByText('Search Criterion').length).toBe(13);
    // And search specifically that fixed length wasn't rendered.
    // There's an extra searchrow from the hooks search dropdown.
    expect(screen.getAllByTestId(`searchrow-${SearchTypesEnum.LENGTH}`).length).toBe(14);
    expect(screen.findAllByTestId(`searchrow-${SearchTypesEnum.FIXED_LENGTH}`).length).toBe(undefined);
  });
});
