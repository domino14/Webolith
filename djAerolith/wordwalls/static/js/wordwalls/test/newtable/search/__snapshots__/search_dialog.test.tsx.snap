// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<SearchDialogContainer /> Static Analysis renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row"
        style="margin-top: 8px;"
      >
        <div
          class="col-sm-8"
        >
          <div>
            <div
              class="row search-row mb-3"
            >
              <div
                class="col-auto d-flex align-items-end pb-3"
              >
                <div>
                  <button
                    class="btn btn-info btn-add-search-row me-1"
                    style="height: 38px;"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="bi bi-plus"
                    />
                  </button>
                  <button
                    class="btn btn-info btn-remove-search-row"
                    style="height: 38px;"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="bi bi-dash"
                    />
                  </button>
                </div>
              </div>
              <div
                class="col-3"
              >
                <div
                  class="mb-3"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-md-12"
                    >
                      <label
                        class="form-label"
                        style="overflow: hidden; white-space: nowrap;"
                      >
                        Search Criterion
                         
                      </label>
                      <select
                        class="form-select"
                        style="min-width: 120px;"
                      >
                        <option
                          data-testid="searchrow-1"
                          value="1"
                        >
                          Word Length
                        </option>
                        <option
                          data-testid="searchrow-2"
                          value="2"
                        >
                          Probability Range
                        </option>
                        <option
                          data-testid="searchrow-8"
                          value="8"
                        >
                          Point Value
                        </option>
                        <option
                          data-testid="searchrow-9"
                          value="9"
                        >
                          Anagram match
                        </option>
                        <option
                          data-testid="searchrow-5"
                          value="5"
                        >
                          Number of Anagrams
                        </option>
                        <option
                          data-testid="searchrow-6"
                          value="6"
                        >
                          Number of Vowels
                        </option>
                        <option
                          data-testid="searchrow-11"
                          value="11"
                        >
                          Not in lexicon
                        </option>
                        <option
                          data-testid="searchrow-17"
                          value="17"
                        >
                          Difficulty range
                        </option>
                        <option
                          data-testid="searchrow-7"
                          value="7"
                        >
                          Has Tags
                        </option>
                        <option
                          data-testid="searchrow-4"
                          value="4"
                        >
                          Probability limit
                        </option>
                        <option
                          data-testid="searchrow-19"
                          value="19"
                        >
                          Words deleted from previous lexicon
                        </option>
                        <option
                          data-testid="searchrow-20"
                          value="20"
                        >
                          Contains Hooks
                        </option>
                        <option
                          data-testid="searchrow-21"
                          value="21"
                        >
                          Definition Contains
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="col"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-6"
                  >
                    <div
                      class="mb-3"
                    >
                      <div
                        class="row"
                      >
                        <div
                          class="col-md-12"
                        >
                          <label
                            class="form-label"
                            style="overflow: hidden; white-space: nowrap;"
                          >
                            Min
                          </label>
                          <input
                            class="form-control"
                            max="15"
                            min="2"
                            type="number"
                            value="7"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-6"
                  >
                    <div
                      class="mb-3"
                    >
                      <div
                        class="row"
                      >
                        <div
                          class="col-md-12"
                        >
                          <label
                            class="form-label"
                            style="overflow: hidden; white-space: nowrap;"
                          >
                            Max
                          </label>
                          <input
                            class="form-control"
                            max="15"
                            min="2"
                            type="number"
                            value="7"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row search-row mb-3"
            >
              <div
                class="col-auto d-flex align-items-end pb-3"
              >
                <div>
                  <button
                    class="btn btn-info btn-add-search-row me-1"
                    style="height: 38px;"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="bi bi-plus"
                    />
                  </button>
                  <button
                    class="btn btn-info btn-remove-search-row"
                    style="height: 38px;"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="bi bi-dash"
                    />
                  </button>
                </div>
              </div>
              <div
                class="col-3"
              >
                <div
                  class="mb-3"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-md-12"
                    >
                      <label
                        class="form-label"
                        style="overflow: hidden; white-space: nowrap;"
                      >
                        Search Criterion
                         
                      </label>
                      <select
                        class="form-select"
                        style="min-width: 120px;"
                      >
                        <option
                          data-testid="searchrow-1"
                          value="1"
                        >
                          Word Length
                        </option>
                        <option
                          data-testid="searchrow-2"
                          value="2"
                        >
                          Probability Range
                        </option>
                        <option
                          data-testid="searchrow-8"
                          value="8"
                        >
                          Point Value
                        </option>
                        <option
                          data-testid="searchrow-9"
                          value="9"
                        >
                          Anagram match
                        </option>
                        <option
                          data-testid="searchrow-5"
                          value="5"
                        >
                          Number of Anagrams
                        </option>
                        <option
                          data-testid="searchrow-6"
                          value="6"
                        >
                          Number of Vowels
                        </option>
                        <option
                          data-testid="searchrow-11"
                          value="11"
                        >
                          Not in lexicon
                        </option>
                        <option
                          data-testid="searchrow-17"
                          value="17"
                        >
                          Difficulty range
                        </option>
                        <option
                          data-testid="searchrow-7"
                          value="7"
                        >
                          Has Tags
                        </option>
                        <option
                          data-testid="searchrow-4"
                          value="4"
                        >
                          Probability limit
                        </option>
                        <option
                          data-testid="searchrow-19"
                          value="19"
                        >
                          Words deleted from previous lexicon
                        </option>
                        <option
                          data-testid="searchrow-20"
                          value="20"
                        >
                          Contains Hooks
                        </option>
                        <option
                          data-testid="searchrow-21"
                          value="21"
                        >
                          Definition Contains
                        </option>
                      </select>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="col"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-6"
                  >
                    <div
                      class="mb-3"
                    >
                      <div
                        class="row"
                      >
                        <div
                          class="col-md-12"
                        >
                          <label
                            class="form-label"
                            style="overflow: hidden; white-space: nowrap;"
                          >
                            Min
                          </label>
                          <input
                            class="form-control"
                            max="70000"
                            min="1"
                            type="number"
                            value="1"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-6"
                  >
                    <div
                      class="mb-3"
                    >
                      <div
                        class="row"
                      >
                        <div
                          class="col-md-12"
                        >
                          <label
                            class="form-label"
                            style="overflow: hidden; white-space: nowrap;"
                          >
                            Max
                          </label>
                          <input
                            class="form-control"
                            max="70000"
                            min="1"
                            type="number"
                            value="200"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            class="btn btn-primary submit-word-search"
            data-bs-dismiss="modal"
            style="margin-top: 0.75em;"
            type="button"
          >
            Play!
          </button>
          <button
            class="btn btn-info"
            data-bs-dismiss="modal"
            style="margin-top: 0.75em; margin-left: 1em;"
            type="button"
          >
            Flashcard
          </button>
        </div>
        <div
          class="col-sm-4"
          style="overflow-y: scroll; height: 400px; margin-top: 32px;"
        >
          <div>
            <button
              class="btn btn-sm btn-info"
              style="font-weight: bold; margin-bottom: 15px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);"
              type="button"
            >
              <span
                style="font-size: 16px; margin-right: 5px;"
              >
                ?
              </span>
               Show Help
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row"
      style="margin-top: 8px;"
    >
      <div
        class="col-sm-8"
      >
        <div>
          <div
            class="row search-row mb-3"
          >
            <div
              class="col-auto d-flex align-items-end pb-3"
            >
              <div>
                <button
                  class="btn btn-info btn-add-search-row me-1"
                  style="height: 38px;"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                    class="bi bi-plus"
                  />
                </button>
                <button
                  class="btn btn-info btn-remove-search-row"
                  style="height: 38px;"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                    class="bi bi-dash"
                  />
                </button>
              </div>
            </div>
            <div
              class="col-3"
            >
              <div
                class="mb-3"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-md-12"
                  >
                    <label
                      class="form-label"
                      style="overflow: hidden; white-space: nowrap;"
                    >
                      Search Criterion
                       
                    </label>
                    <select
                      class="form-select"
                      style="min-width: 120px;"
                    >
                      <option
                        data-testid="searchrow-1"
                        value="1"
                      >
                        Word Length
                      </option>
                      <option
                        data-testid="searchrow-2"
                        value="2"
                      >
                        Probability Range
                      </option>
                      <option
                        data-testid="searchrow-8"
                        value="8"
                      >
                        Point Value
                      </option>
                      <option
                        data-testid="searchrow-9"
                        value="9"
                      >
                        Anagram match
                      </option>
                      <option
                        data-testid="searchrow-5"
                        value="5"
                      >
                        Number of Anagrams
                      </option>
                      <option
                        data-testid="searchrow-6"
                        value="6"
                      >
                        Number of Vowels
                      </option>
                      <option
                        data-testid="searchrow-11"
                        value="11"
                      >
                        Not in lexicon
                      </option>
                      <option
                        data-testid="searchrow-17"
                        value="17"
                      >
                        Difficulty range
                      </option>
                      <option
                        data-testid="searchrow-7"
                        value="7"
                      >
                        Has Tags
                      </option>
                      <option
                        data-testid="searchrow-4"
                        value="4"
                      >
                        Probability limit
                      </option>
                      <option
                        data-testid="searchrow-19"
                        value="19"
                      >
                        Words deleted from previous lexicon
                      </option>
                      <option
                        data-testid="searchrow-20"
                        value="20"
                      >
                        Contains Hooks
                      </option>
                      <option
                        data-testid="searchrow-21"
                        value="21"
                      >
                        Definition Contains
                      </option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="row"
              >
                <div
                  class="col-6"
                >
                  <div
                    class="mb-3"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col-md-12"
                      >
                        <label
                          class="form-label"
                          style="overflow: hidden; white-space: nowrap;"
                        >
                          Min
                        </label>
                        <input
                          class="form-control"
                          max="15"
                          min="2"
                          type="number"
                          value="7"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="col-6"
                >
                  <div
                    class="mb-3"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col-md-12"
                      >
                        <label
                          class="form-label"
                          style="overflow: hidden; white-space: nowrap;"
                        >
                          Max
                        </label>
                        <input
                          class="form-control"
                          max="15"
                          min="2"
                          type="number"
                          value="7"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="row search-row mb-3"
          >
            <div
              class="col-auto d-flex align-items-end pb-3"
            >
              <div>
                <button
                  class="btn btn-info btn-add-search-row me-1"
                  style="height: 38px;"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                    class="bi bi-plus"
                  />
                </button>
                <button
                  class="btn btn-info btn-remove-search-row"
                  style="height: 38px;"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                    class="bi bi-dash"
                  />
                </button>
              </div>
            </div>
            <div
              class="col-3"
            >
              <div
                class="mb-3"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-md-12"
                  >
                    <label
                      class="form-label"
                      style="overflow: hidden; white-space: nowrap;"
                    >
                      Search Criterion
                       
                    </label>
                    <select
                      class="form-select"
                      style="min-width: 120px;"
                    >
                      <option
                        data-testid="searchrow-1"
                        value="1"
                      >
                        Word Length
                      </option>
                      <option
                        data-testid="searchrow-2"
                        value="2"
                      >
                        Probability Range
                      </option>
                      <option
                        data-testid="searchrow-8"
                        value="8"
                      >
                        Point Value
                      </option>
                      <option
                        data-testid="searchrow-9"
                        value="9"
                      >
                        Anagram match
                      </option>
                      <option
                        data-testid="searchrow-5"
                        value="5"
                      >
                        Number of Anagrams
                      </option>
                      <option
                        data-testid="searchrow-6"
                        value="6"
                      >
                        Number of Vowels
                      </option>
                      <option
                        data-testid="searchrow-11"
                        value="11"
                      >
                        Not in lexicon
                      </option>
                      <option
                        data-testid="searchrow-17"
                        value="17"
                      >
                        Difficulty range
                      </option>
                      <option
                        data-testid="searchrow-7"
                        value="7"
                      >
                        Has Tags
                      </option>
                      <option
                        data-testid="searchrow-4"
                        value="4"
                      >
                        Probability limit
                      </option>
                      <option
                        data-testid="searchrow-19"
                        value="19"
                      >
                        Words deleted from previous lexicon
                      </option>
                      <option
                        data-testid="searchrow-20"
                        value="20"
                      >
                        Contains Hooks
                      </option>
                      <option
                        data-testid="searchrow-21"
                        value="21"
                      >
                        Definition Contains
                      </option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="row"
              >
                <div
                  class="col-6"
                >
                  <div
                    class="mb-3"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col-md-12"
                      >
                        <label
                          class="form-label"
                          style="overflow: hidden; white-space: nowrap;"
                        >
                          Min
                        </label>
                        <input
                          class="form-control"
                          max="70000"
                          min="1"
                          type="number"
                          value="1"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="col-6"
                >
                  <div
                    class="mb-3"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col-md-12"
                      >
                        <label
                          class="form-label"
                          style="overflow: hidden; white-space: nowrap;"
                        >
                          Max
                        </label>
                        <input
                          class="form-control"
                          max="70000"
                          min="1"
                          type="number"
                          value="200"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          class="btn btn-primary submit-word-search"
          data-bs-dismiss="modal"
          style="margin-top: 0.75em;"
          type="button"
        >
          Play!
        </button>
        <button
          class="btn btn-info"
          data-bs-dismiss="modal"
          style="margin-top: 0.75em; margin-left: 1em;"
          type="button"
        >
          Flashcard
        </button>
      </div>
      <div
        class="col-sm-4"
        style="overflow-y: scroll; height: 400px; margin-top: 32px;"
      >
        <div>
          <button
            class="btn btn-sm btn-info"
            style="font-weight: bold; margin-bottom: 15px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);"
            type="button"
          >
            <span
              style="font-size: 16px; margin-right: 5px;"
            >
              ?
            </span>
             Show Help
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
